actionpack/lib/action_controller/metal/request_forgery_protection.rb

```diff
@@ -253,6 +300,47 @@ def handle_unverified_request
       end
     end
 
+    module TokenStorageMethods
+      class Base
+        attr_reader :controller
+
+        def initialize(controller)
+          @controller = controller
+        end
+
+        private
+          def generate_csrf_token # :nodoc:
+            if controller.urlsafe_csrf_tokens
+              SecureRandom.urlsafe_base64(AUTHENTICITY_TOKEN_LENGTH)
+            else
+              SecureRandom.base64(AUTHENTICITY_TOKEN_LENGTH)
+            end
+          end
+      end
+
+      class SessionStorage < Base
+        def initialize(controller)
+          super
+        end
+
+        def retrieve_token
+          controller.request.session[:_csrf_token] ||= generate_csrf_token
+        end
+      end
+
+      class CookieStorage < Base
+        attr_reader :cookie
+
+        def initialize(controller)
+          super
+        end
+
+        def retrieve_token
+          controller.request.cookie_jar.encrypted[controller.csrf_token_cookie] ||= generate_csrf_token
```

Comment:
Not that it's a big deal, but do we even need to encrypt the cookie? 

Also I think it would be best to set it as `httponly`, no?

Reply:
I went with an encrypted cookie because my understanding is that by default, the rails session is actually an encrypted cookie, so I wanted to keep the encryption the same when moving from session to standalone cookie.  If there is any way for an attacker to access the cookie, encrypting it will ensure that to token is secure.  That said, I'm not claiming to be an expert on web security, so it's possible we don't need to encrypt it.

With respect to `httponly`, I was thinking the same thing while getting ready for work this morning.  I'll make that change and take a look at the configuration tests which seem to be failing.

Reply:
> so it's possible we don't need to encrypt it

Pretty sure we don't as the token is in clear in the HTML header anyway.

Reply:
> Pretty sure we don't as the token is in clear in the HTML header anyway.

If an attacker can access an unencrypted cookie (somehow), couldn't they potentially use the value stored in the cookie to create a form with that same token, which would then allow a CSRF attack to succeed?

Reply:
Yes, but the protection rely on the attacker not being able to get the cookie. The cookie is supposed to be `HttpOnly` and `SameSite`. And as I said, the token is also present on in the HTML header, so really an unencrypted cookie really isn't any less secure.

Reply:
See: https://en.wikipedia.org/wiki/Cross-site_request_forgery#Prevention

Reply:
Ok.  You've convinced me.  Thanks.